Fix Concave, Skip and Clip Collision Build Bugs

RhapsodyInGeek opened the pull request:
Most of Qodot's MapData is passed by Span, a data type that references another existing value. Modifying the Span value will modify the original value as well.

Qodot generates Mesh and Collision Shape data running 2 passes of SurfaceGatherer.Run() on each Entity. During each Run(), SurfaceGatherer loops through each Brush and Face of an SolidEntity, checking if the current Brush or Face it's on should be skipped. It then loops through each Vertex of the current Face. If the current SolidEntity is an Entity or Group type, it subtracts the center position from the current Vertex. Then finally it adds the Vertex to the Surface Output.

The problem is that it loops through the Vertices as a Span type, which is passed by reference. Therefore, every time the Entity's Center position is subtracted from the Vertex, it modifies the original value. For Skip textures, it only ends subtracting from this value once because they are filtered out during the Mesh Generation Run. Meanwhile, all other texture types are subtracted twice, causing the sets of vertices to be offset.

The fix is to copy the Span's current FaceVertex (vertexSpan[v]) to a new local FaceVertex struct, then modify the copied value and insert that into the Output array, leaving the original MapData intact.

With this change, there is no longer a need to add the entity.center to the vertex position in qodot_map.gd's build_entity_collision_shapes() function, nor is there any need to comment out GeoGenerator's entity.center += brush.center line.

FilterBrush is also moved back to the start of the Brush loop as in the original Qodot for 3.5. This seems to solve the crashes related to Clip texture filtering.

Tested both on the default qodot.fgd with worldspawn and on a custom FGD with special SolidEntity classes.

EMBYRDEV commented:
Thank you for this! Have you got a good test project I can use to examine the before/after output of this change. WIll check this out as soon as possible!

RhapsodyInGeek commented:
I don't yet but I'll fix one up this morning.

RhapsodyInGeek commented:
Here's a minimum reproduction example using the example qodot.fgd. It uses Skip and Clip on the worldspawn, Skip on a mover SolidEntity, and Clip on a Trigger.

Current Qodot design has Skip restricted to removing faces (though it is capable of removing the faces of something entirely Skip textured) while Clip is restricted to only removing faces if the entire brush (not necessarily entity) is textured with Clip.

This feels a bit redundant so I've been thinking of other design ideas to distinguish Clip and Skip (Ember already made a proposal awhile ago https://github.com/QodotPlugin/Qodot/issues/72 but I think there's a simpler solution to get the same result) but that's outside of the scope of this fix.

Here's the minimum repro project, hope it uploads!
[QodotTest.zip](https://github.com/QodotPlugin/Qodot/files/12706363/QodotTest.zip)

RhapsodyInGeek commented:
Fixed Concave Collision Building. Turns out it was a small oversight in [SurfaceGatherer's Run() method](https://github.com/QodotPlugin/Qodot/blob/main/addons/qodot/src/core/SurfaceGatherer.cs#L34). I've also updated the `build_entity_collision_shapes` function in [qodot_map.gd](https://github.com/QodotPlugin/Qodot/blob/main/addons/qodot/src/nodes/qodot_map.gd#L767), since with the Skip and Clip fix it's no longer necessary, just like the convex section of the code.

dcealopez commented:
I tested these changes and there seems to be an issue where the collisions are displaced for solid entities with collision type set to convex in the FGD.

![imagen](https://github.com/QodotPlugin/Qodot/assets/22688330/fecf1336-4fbc-4c0c-b3ff-b638c891f6aa)

Works fine with concave collisions, no issues there.

RhapsodyInGeek commented:
Can you share your Map file, because I'm not running into any problems with this code across 3 different projects.

dcealopez commented:
> Can you share your project, because I'm not running into any problems with this code across 3 different projects.

Nevermind. It was a mistake on my part when cloning the changes of this PR.

DeerTears commented:
Tested, looks great!
